analytics/tracker.go

```diff
@@ -26,110 +78,166 @@ type Tracker interface {
+	if !t.stateChecker.Enabled() {
+		return
+	}
+
+	var eventName string
+	var extraProperties analytics.Properties
+
 	switch results.Status {
 	case models.StepRunStatusCodeSuccess:
-		t.tracker.Enqueue("step_finished", properties, analytics.Properties{"status": "successful"})
+		eventName = stepFinishedEventName
+		extraProperties = analytics.Properties{statusProperty: successfulValue}
 		break
 	case models.StepRunStatusCodeFailed, models.StepRunStatusCodeFailedSkippable:
-		failedProperties := analytics.Properties{"status": "failed"}
-		failedProperties.AppendIfNotEmpty("error_message", results.ErrorStr)
-		t.tracker.Enqueue("step_finished", properties, failedProperties)
+		eventName = stepFinishedEventName
+		extraProperties = analytics.Properties{statusProperty: failedValue}
+		extraProperties.AppendIfNotEmpty(errorMessageProperty, results.ErrorStr)
 		break
 	case models.StepRunStatusCodePreparationFailed:
-		failedProperties := prepareStartProperties(results.StepInfo)
-		failedProperties.AppendIfNotEmpty("error_message", results.ErrorStr)
-		t.tracker.Enqueue("step_preparation_failed", properties, failedProperties)
+		eventName = stepPreparationFailedEventName
+		extraProperties = prepareStartProperties(results.StepInfo)
+		extraProperties.AppendIfNotEmpty("error_message", results.ErrorStr)
 	case models.StepRunStatusCodeSkipped, models.StepRunStatusCodeSkippedWithRunIf:
-		startProperties := prepareStartProperties(results.StepInfo)
+		eventName = stepSkippedEventName
+		extraProperties = prepareStartProperties(results.StepInfo)
 		if results.Status == models.StepRunStatusCodeSkipped {
-			startProperties["reason"] = "build_failed"
+			extraProperties[reasonProperty] = buildFailedValue
 		} else {
-			startProperties["reason"] = "run_if"
+			extraProperties[reasonProperty] = runIfValue
 		}
-		t.tracker.Enqueue("step_skipped", properties, startProperties)
 	default:
-		panic("Unknown step status code")
+		panic(fmt.Sprintf("Unknown step status code: %d", results.Status))
 	}
+
+	t.tracker.Enqueue(eventName, properties, extraProperties)
 }
 
-func prepareStartProperties(infoModel stepmanModels.StepInfoModel) analytics.Properties {
-	properties := analytics.Properties{}
-	properties.AppendIfNotEmpty("step_id", infoModel.ID)
-	properties.AppendIfNotEmpty("step_title", pointers.StringWithDefault(infoModel.Step.Title, ""))
-	properties.AppendIfNotEmpty("step_version", infoModel.Version)
-	properties.AppendIfNotEmpty("step_source", pointers.StringWithDefault(infoModel.Step.SourceCodeURL, ""))
-	properties["skippable"] = pointers.BoolWithDefault(infoModel.Step.IsSkippable, false)
-	return properties
+// SendCLIWarning ...
+func (t tracker) SendCLIWarning(message string) {
+	if !t.stateChecker.Enabled() {
+		return
+	}
+
+	t.tracker.Enqueue(cliWarningEventName, analytics.Properties{messageProperty: message})
 }
 
 // Wait ...
 func (t tracker) Wait() {
+	if !t.stateChecker.Enabled() {
```

Comment:
I think in this case we should let the call through `t.tracker.Wait()` even if analytics got disabled on the fly. I'm not confident that skipping the `Wait()` call won't cause a deadlock or memory leaks.